src/wp-includes/canonical.php

```diff
@@ -972,10 +972,36 @@ function redirect_guess_404_permalink() {
 		 */
 		$strict_guess = apply_filters( 'strict_redirect_guess_404_permalink', false );
 
+		$query_args = array(
+			'post_status'            => $publicly_viewable_statuses,
+			'posts_per_page'         => 1,
+			'no_found_rows'          => true,
+			'ignore_sticky_posts'    => true,
+			'update_post_meta_cache' => false,
+			'update_post_term_cache' => false,
+			'fields'                 => 'ids',
+		);
+
+		// Handle strict vs. loose post_name matching.
 		if ( $strict_guess ) {
-			$where = $wpdb->prepare( 'post_name = %s', get_query_var( 'name' ) );
+			$query_args['name'] = get_query_var( 'name' );
 		} else {
-			$where = $wpdb->prepare( 'post_name LIKE %s', $wpdb->esc_like( get_query_var( 'name' ) ) . '%' );
+			// For loose matching (LIKE), we'll use a posts_where filter.
+			$post_name_for_filter = get_query_var( 'name' );
+
+			// Store the filter callback so we can remove it later.
+			$post_name_where_filter = static function ( $where, $query ) use ( $post_name_for_filter, $wpdb ) {
+				// Only apply to our specific query.
+				if ( isset( $query->query_vars['redirect_guess_404'] ) ) {
+					$where .= $wpdb->prepare( " AND {$wpdb->posts}.post_name LIKE %s", $wpdb->esc_like( $post_name_for_filter ) . '%' );
+				}
+				return $where;
+			};
+
+			add_filter( 'posts_where', $post_name_where_filter, 10, 2 );
+
+			// Mark this query so our filter knows to apply the LIKE clause.
+			$query_args['redirect_guess_404'] = true;
```

Comment:
```suggestion
			$query_args['s'] = get_query_var( 'name' );
			$query_args['search_columns'] = array( 'post_name' );
```

Currently, WP_Query does not support search column of post name, but it is a very small tweak to make this happen.